microservices/object-identifer-strategy.md

```diff
@@ -0,0 +1,140 @@
+
+The following rules must be adhered to during UUID generation:
+
+* UUIDs must be generated using the version 4 (random) algorithm
+* UUIDs must be generated within the application code (examples are given below), not within the database. This avoids the problem whereby every time you need to insert an entity, you need the database to give you its primary key i.e. the entities in your code are incomplete until you talk to the database
+* UUIDs must be stored in lowercase using the 8-4-4-4-12 format e.g: `123e4567-e89b-12d3-a456-426655440000`
+
+## URNs
+The use of [Uniform Resource Names](https://en.wikipedia.org/wiki/Uniform_Resource_Name) (URNs) was considered, but rejected. URNs allow the identity of a RESTful resource (object) to be separated from its location. A good analogy is to think of how we generally refer to people by their name rather than their home address, because the latter can change. Using URNs requires the use of a URN resolver, which is like an address book that looks up the URN it is given and produces the [Uniform Resource Identifier](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) (URI) where that resource can be found. The use of a URN resolver increases network traffic because every reference to a resource requires an additional call to look up its location.
+
+There is currently no clear requirement for the SDC platform that justifies the additional development and operational complexity that would be required to fully implement URNs. If a requirement emerges later it would be easy to add to the strategy proposed here by prefixing all existing UUIDs with a URN namespace identifier (NID) such as `uk.gov.ons.<type>`
+
+## Worked Example
+The example below illustrate how the use of UUIDs to identify objects would work in practice. Rather than get bogged down in the detail the important thing is to note how the UUIDs are passed around between the three collaborating microservices.
+
+For this example the Collection Instrument service needs to load an offline collection instrument (spreadsheet) into the system so it's associated with the correct reporting unit. The RAS back office UI makes a request to the Survey service to get the list of known surveys so it can present their names to the user:
+
+`GET https://surveysvc/surveys`
+
+```json 
+[{
+    "id": "cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87",
+    "shortName": "BRES"
+}]
+```
+
+The user selects the survey applicable to the collection instrument they want to load e.g. BRES. At this point the application has the unique identifier for the BRES survey object and can ask the Collection Exercise service for the list of collection exercises for that survey:
+
+`GET https://collectionexercisesvc/collectionexercises/survey/cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87`
```

Comment:
This is probably outside of the context of this standard, but why not have `survey` as a query parameter on `collectionexercises` e.g. `GET https://collectionexercisesvc/collectionexercises?survey=cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87` or a more generic `GET https://collectionexercisesvc/collectionexercises?filter=survey:cb0711c3-0ac8-41d3-ae0e-567e5ea1ef87`

Reply:
It's certainly a valid approach, but I would regard survey as a secondary RESTful resource (collection exercise is the primary resource) in this context. I tend to keep query parameters for other concerns such as filtering, ordering and pagination. I know there's an argument to be made that you're filtering by survey here...